### PR TITLE
Frontmatter typing

### PR DESCRIPTION
**Pull request by wackerow**
## (Builds off Roadmap/Upgrade layouts in #38)
This built off the other PR open (branch `roadmap-layout`) so I targeted that branch to make the diff for these changes easier to see.

## Description
Implements typing schema for all front matter properties used throughout all layouts.

Placed primarily within `interfaces.ts`, the schema starts with a `SharedFrontmatter` interface, and then individual interfaces are built via extension to match the front matter schema used in each layout, with a naming pattern of `LayoutNameFrontmatter`.

A general `Frontmatter` type is declared in `types.ts` as a union type of all individual interfaces, used in cases before the exact schema needed is known (i.e., in `getContentBySlug` used in `[...slug].tsx` and `[...tutorial].tsx`)

Renamed `PageContent` interface to `MdPageContent` to increase specificity in a shared `interfaces.ts` file... in this, updated the `frontmatter` field to use the union type of `Frontmatter`.

In each `layouts/Layout.tsx` file, using `MdPageContent` extended in it's prop typing, the `frontmatter` field is then overwritten with specifically which frontmatter interface should be used within the layout, ie `frontmatter: UseCasesFrontmatter` in `layouts/UseCases.tsx` IProps

Note: Docs layout is not yet implemented, nor is Staking, but these have both been prepared as well.

### CODE REVIEW

File: src/lib/utils/md.ts [hunk 1]
```diff
@@ -91,9 +92,9 @@ export const getContentBySlug = (slug: string) => {
   const realSlug = `${slug}/index.md`
   const fullPath = join(CURRENT_CONTENT_DIR, realSlug)
   const fileContents = fs.readFileSync(fullPath, "utf8")
-  const { data: frontmatter, content } = matter(fileContents)
-
-  const items: PageContent = {
+  const { data, content } = matter(fileContents)
+  const frontmatter = data as Frontmatter
```
Comment (wackerow) [inline src/lib/utils/md.ts:115]: Typing as union type `Frontmatter` at this point, which will allow front matter formatted for _any_ of the acceptable layouts.

File: src/lib/interfaces.ts [hunk 2]
```diff
@@ -56,26 +56,90 @@ export interface ToCItem {
...
+  postMergeBannerTranslation?: string
+  hideEditButton?: boolean
+}
+
+export interface MdPageContent {
+  slug: string
+  content: string
+  frontmatter: Frontmatter
+  tocItems: Array<ToCItem>
+  lastUpdatedDate?: string
```
Comment (wackerow) [inline src/lib/interfaces.ts]: Union type `Frontmatter` used here to accept all types of frontmatter with page content; cast as specific layout frontmatter type within layout file to enforce layout-specific schema.

File: src/layouts/Roadmap.tsx [hunk 3]
```diff
@@ -103,7 +103,9 @@ export const roadmapComponents = {
   // MeetupList,
 }
 
-interface IProps extends PageContent, ChildOnlyProp {}
+interface IProps extends MdPageContent, ChildOnlyProp {
+  frontmatter: RoadmapFrontmatter
+}
 export const RoadmapLayout: React.FC<IProps> = ({ children, frontmatter, slug, tocItems }) => {
```
Comment (wackerow) [inline src/layouts/Roadmap.tsx]: `MdPageContent` contains a `frontmatter` property typed as `Frontmatter` which is a union type of any possible frontmatter schema... Inside the layout here we can then cast it as the layout-specific frontmatter schema needed to enforce typing

(pattern repeated on all layouts)

File: src/layouts/Static.tsx [hunk 4]
```diff
@@ -171,6 +172,9 @@ export const staticComponents = {
   YouTube,
 }
 
+interface IProps extends ChildOnlyProp, MdPageContent {
+  frontmatter: StaticFrontmatter
+}
 export const StaticLayout = ({
```
Comment (pettinarip) [inline src/layouts/Static.tsx]:
```suggestion
export const StaticLayout: React.FC<IProps> = ({
```

File: src/lib/interfaces.ts [hunk 5]
```diff
@@ -56,26 +56,90 @@ export interface ToCItem {
...
+  skill: string
+  published: string
+  address?: string
+  postMergeBannerTranslation?: string
+  hideEditButton?: boolean
+}
+
+export interface MdPageContent {
+  slug: string
+  content: string
```
Comment (pettinarip) [inline src/lib/interfaces.ts:147]: What is `content`? I don't think we are passing that prop. The `children` would be our "content" now.

**Review by pettinarip — APPROVED** [review]
Nice! LGTM

Left 2 observations but not blockers.

File: src/lib/interfaces.ts (hunk 5 above)
Comment (wackerow) [reply src/lib/interfaces.ts:147]: image

The body of the markdown is assigned to `content` first here in `getContentBySlug`, called during `getStaticProps` in `[...slug]`, and then that `content` is serialized into `mdxSource`:

image

And then passed to next MDX remote:

image

**Pull request merged**